pyasdf/asdf_data_set.py

```diff
@@ -1451,18 +1456,123 @@ def split(container, count):
+        """
+
+        def create_dataset_params(sta_info):
+
+            # Split the path
+            tag_path = sta_info["path"].strip("/").split("/")
+
+            for path in tag_path:
+                # Assert each path piece.
+                tag_pattern = r"^[a-zA-Z0-9][a-zA-Z0-9_]*[a-zA-Z0-9]$"
+                if re.match(tag_pattern, path) is None:
+                    raise ASDFValueError(
+                        "Tag name '{name}' is invalid. It must validate "
+                        "against the regular expression '{pattern}'.".format(
+                            name=path, pattern=tag_pattern))
+
+            info = {
+                "data_name": sta_info["path"],
+                "dataset_creation_params": {
+                    "name": sta_info["path"],
+                    "shape": sta_info["object"].shape,
+                    "dtype": sta_info["object"].dtype,
+                    "compression": self.__compression[0],
+                    "compression_opts": self.__compression[1],
+                    "shuffle": self.__shuffle,
+                    "fletcher32": False,
+                    "maxshape": tuple([None] * len(sta_info["object"].shape))
+                },
+                "dataset_attrs": sta_info["parameters"],
+            }
+            return info
+
+        meta_list = []
+        for _sta, _sta_info in results.iteritems():
+            for _chan_info in _sta_info:
+                _meta = create_dataset_params(_chan_info)
+                meta_list.append(_meta)
+
+        gathered_meta = self.mpi.comm.gather(meta_list, root=0)
+        all_meta = []
         if self.mpi.rank == 0:
-            for result in gathered_results:
-                results.update(result)
+            for _meta in gathered_meta:
+                all_meta.extend(_meta)
+
+        all_meta = self.mpi.comm.bcast(all_meta, root=0)
 
         # Likely not necessary as the gather two line above implies a
         # barrier but better be safe than sorry.
         self.mpi.comm.barrier()
 
-        return results
+        return all_meta
+
+    def _write_hdf5_dataset_meta_mpi(self, group, results):
+        """
+        Write meta information of auxiliary data into hdf5,
+        collectivelly
+        """
+        # collect necessary information without data
+        collect_info = self._collect_hdf5_meta_mpi(results)
+
+        for _sta_info in collect_info:
+            ds = group.create_dataset(**_sta_info["dataset_creation_params"])
+
+            for key, value in _sta_info["dataset_attrs"].iteritems():
+                ds.attrs[key] = value
+
+    def _write_hdf5_dataset_data_mpi(self, group, results):
+        """
+        Write auxiliary data into hdf5, indepandantly
+        """
+        for _sta, _sta_info in results.iteritems():
+            for _chan_info in _sta_info:
+                path = _chan_info["path"]
+                adj_data = _chan_info["object"]
+                group[path][:] = adj_data
```

Comment:
Find it a bit odd...Maybe you know why it is so.

First, I set the `adjoint_source.shape = [NPTS, 2]`. `adjoint_source[:, 0]` is time array, and the other one is displacement array of adjoint source. I found when writing this piece into hdf5 is really slow at:
```
      group[path][:] = adj_data
```
For my test case, the size of hdf5 file is 600MB.

However, when I set `adjoint_source.shape = [NPTS, ]`. Basically, just keep track of the displacement. For the time, I just simple add another attribute called "time_offset". Together with the attribute of "dt", we could reconstruct the time array without any issue. This operation will reduce the datasize by the factor of 2, which gives me 300MB data.

However, I feel the running time reduction could be a factor of 10. Do you know why?